Lib/datetime.py

```diff
@@ -1502,32 +1554,36 @@ def dst(self):
         return offset
 
     def replace(self, hour=None, minute=None, second=None, microsecond=None,
-                tzinfo=True, *, fold=None):
+                tzinfo=True, *, nanosecond=None, fold=None):
         """Return a new time with new values for the specified fields."""
         if hour is None:
             hour = self.hour
         if minute is None:
             minute = self.minute
         if second is None:
             second = self.second
-        if microsecond is None:
-            microsecond = self.microsecond
+        if nanosecond is None:
```

Comment:
As much as I don't like gratuitous introduction of floating point numbers, I think in this case it is best to allow fractional number of microseconds in the datetime constructor.  (Accepting numbers.Rational). We can still add nanoseconds keyword to `timedelta` where all keywords already support passing floats.

Reply:
We did discuss this at the CPython core dev sprint, in the `#stdlib` room. A relevant excerpt from the conversation (@gpshead and @gvanrossum, please let me know if you'd like me to remove the transcript and replace it with a summary):

> [2020-10-23 15:40:18] pganssle: Yeah. The API is always going to be a little awkard, particularly because of the positional arguments (which we can't change), but I think "can't specify both at once" is probably the safest.
> [2020-10-23 15:41:36] pganssle: It might also make sense to accept a float for `microsecond` as an alternate way to specify nanoseconds, but that could always be added later if we decide it's desirable.
> [2020-10-23 21:51:19] gpshead (Greg): accepting a float for microsecond on *input* is an option.  a double has enough digits to be precise to nanoseconds within a 0.0 - 999_999_999.999 range.  i'd just consider it a convenience though rather than a good primary api.
> [2020-10-23 21:54:41] gvanrossum: Hm. To me, "float" time implies "seconds". Would hate to lose that for such a minor convenience.
> [2020-10-23 21:55:49] gvanrossum: (It's not an important principle, but a handy debugging heuristic -- "hey, these times are floats; they probably meant seconds")
> [2020-10-23 21:57:25] gpshead (Greg): yeah, it'd be awkward to specify a fractional component of something as a range restricted float.
> [2020-10-23 22:02:05] pganssle: I just meant that you could pass `microsecond=100.145` to be the same as `microsecond=None, nanosecond=100_145`.
> [2020-10-23 22:02:29] pganssle: The return value of `.microsecond` would still be `100` and `.nanosecond` would be `100145`.
> [2020-10-23 22:03:52] pganssle: The biggest problem I have with that as a convenience API would be that it's mostly unnecessary and also it's got a non-obvious precision truncation going on. `datetime(2020, 1, 1, microsecond=100.145)` would be the same as `datetime(2020, 1, 1, microsecond=100.14577893940)`, even if those are two distinct floats.
> [2020-10-23 22:18:17] gvanrossum: Still -1.0 on that.
> [2020-10-23 22:21:04] gpshead (Greg): yeah agreed now that i see it. (-1)

It seems that the rough consensus on this is that there's nowhere to go from here that _isn't_ awkward or problematic in some way, but people prefer `microsecond` and `nanosecond` to be mutually exclusive ints.

Reply:
GvR is right (no surprise!) - we should allow float (rational) seconds instead of microseconds and disallow microseconds in this case.  `time(12, 0, 59.123456789)` looks much better than `time(12, 0, 59, 123.456789)`.

I would even argue that positional microsecond argument can be deprecated or at least discouraged in the future in favor of rational seconds).

Reply:
I think it _might_ be more annoying than it's worth to deprecate `microsecond` here (and in the C API!), because there's probably a lot of code out there specifying microseconds that would need to be changed for something so trivial. Otherwise I agree with the path forward: accept floats for `second`.

The remaining questions:

1. Do we still add a `nanosecond` kwarg? Clearly it would be similarly discouraged _but_ people don't usually pay attention to us, but they _will_ notice that there's a `microsecond` but no `nanosecond`, and they won't understand why. I definitely foresee adding nanosecond support without a `nanosecond` kwarg to be greeted with complaints and confusion.
2. Do we still allow floats/rationals on `microsecond`, or are we going to say "fractional seconds *or* specify `microsecond` *or* specify `nanosecond`, but only one of the three"? I'm OK with either but leaning towards the second.
3. If we do this, will it / should it affect the exostomg C API? I'm thinking no, we should simply add an additional / additional constructors (API to be specified later, but probably along the same lines — one that takes a `double` for `second` and/or one that takes a `Nanosecond` positional argument).